tsl/src/planner.c

```diff
@@ -231,6 +238,46 @@ tsl_create_distributed_insert_path(PlannerInfo *root, ModifyTablePath *mtpath, I
 
 			table_close(rel, AccessShareLock);
 		}
+		else /* check if it's a INSERT ... SELECT case involving dist hypertables */
+		{
+			RangeTblEntry *select_rte = NULL;
+			ListCell *l;
+			RangeTblEntry *rte = planner_rt_fetch(hypertable_rti, root);
+			bool distributed = false;
+
+			/* if src hypertable is distributed then only further checks are needed */
+			if (ts_rte_is_hypertable(rte, &distributed) && distributed)
+			{
+				/* check if it's indeed INSERT .. SELECT */
+				foreach (l, root->parse->rtable)
+				{
+					rte = (RangeTblEntry *) lfirst(l);
+
+					if (rte->rtekind == RTE_SUBQUERY)
+					{
+						select_rte = rte;
+						break;
+					}
+				}
+
+				/* check if the SELECT is referring to a distributed hypertable */
+				if (select_rte)
+				{
+					Query *subquery = select_rte->subquery;
+
+					foreach (l, subquery->rtable)
+					{
+						rte = (RangeTblEntry *) lfirst(l);
+
+						if (ts_rte_is_hypertable(rte, &distributed) && distributed)
```

Comment:
This doesn't recurse, does it? We can construct an arbitrarily deep nesting of RTE_SUBQUERY's and put a distributed table at the bottom, and it'll fail. Probably we have to recurse. To construct an example query, we can prevent the subqueries being pulled up by putting a volatile function in their targetlist, such as `random()`.

Reply:
Ah, right. So I fixed this by checking entries in the `root->simple_rte_array` array starting from the index beyond that of the source hypertable.